# HG changeset patch
# User Rik <rik@octave.org>
# Date 1426317614 25200
#      Sat Mar 14 00:20:14 2015 -0700
# Node ID b5479bbc3727acc2bf352fd79602e2dcea960f11
# Parent  af1a024b3344230343e2c5b1de9e37228c04a013
NEWS: Update NEWS file for 4.0 release.

* NEWS: Add new features, new functions, and deprecated functions to file.
Re-order items for clarity of presentation.

diff --git a/NEWS b/NEWS
--- a/NEWS
+++ b/NEWS
@@ -1,11 +1,30 @@
 Summary of important user-visible changes for version 4.0:
 ---------------------------------------------------------
 
+ ** A graphical user interface is now the default when running Octave
+    interactively.  The start-up option --no-gui will run the familiar
+    command line interface, and still allows use of the GUI dialogs and
+    qt plotting toolkit.  The option --no-gui-libs runs a minimalist
+    command line interface that does not link with the Qt libraries and
+    uses the fltk toolkit for plotting.
+
+ ** Octave now uses OpenGL graphics with Qt widgets by default.  If OpenGL
+    libraries are not available when Octave is built, gnuplot is used.
+    You may choose to use the fltk or gnuplot toolkit for graphics by
+    executing the command
+
+      graphics_toolkit ("fltk")
+        OR
+      graphics_toolkit ("gnuplot")
+
+    Adding such a command to your ~/.octaverc file will set the default
+    for each session.
+
  ** A new syntax for object oriented programming termed classdef has been
     introduced.  See the manual for more extensive documentation of the
     classdef interface.
 
     New keywords:
 
       classdef      endclassdef
       enumeration   endenumeration
@@ -32,28 +51,16 @@ Summary of important user-visible change
     which control whether the optimizations are used (default), or whether
     the data types are stored as full matrices.
 
     disable_range   disable_diagonal_matrix   disable_permutation_matrix
 
     All three optimizations are disabled if Octave is started with the
     --braindead command line option.
 
- ** For compatibility with Matlab, the "backtrace" warning option is now
-    enabled by default.
-
- ** For compatibility with Matlab, the "ismatrix(x)" function now only checks
-    the dimension of "x". The old behaviour of "ismatrix" is obtained by
-    "isnumeric (x) || islogical (x) || ischar (x)".
-
- ** The nextpow2 function behaviour has been changed for vector inputs.
-    Instead of computing `nextpow2 (length (x))', it will now compute
-    nextpow2() for each element of the input.  Not only is this change
-    Matlab compatible, it also prevents bugs for "vectors" of length 1.
-
  ** The preference
 
       do_braindead_shortcircuit_evaluation
 
     is now enabled by default.
 
  ** The preference
 
@@ -61,16 +68,28 @@ Summary of important user-visible change
 
     is now enabled by default and the warning ID
 
       Octave:noninteger-range-as-index
 
     is now set to "on" by default instead of "error" by default and "on"
     for --traditional.
 
+ ** The "backtrace" warning option is now enabled by default.  This change
+    was made for Matlab compatability.
+
+ ** For compatibility with Matlab, the "ismatrix (x)" function now only checks
+    the dimension of "x".  The old behaviour of "ismatrix" is obtained by
+    "isnumeric (x) || islogical (x) || ischar (x)".
+
+ ** The nextpow2 function behaviour has been changed for vector inputs.
+    Instead of computing `nextpow2 (length (x))', it will now compute
+    nextpow2 for each element of the input.  This change is Matlab compatible,
+    and also prevents bugs for "vectors" of length 1.
+
  ** polyeig now returns a row vector of eigenvalues rather than a matrix
     with the eigenvalues on the diagonal.  This change was made for Matlab
     compatibility.
 
  ** Interpolation function changes for Matlab compatibility
 
     The interpolation method 'cubic' is now equivalent to 'pchip'
     for interp1, interp2, and interp3.  Previously, 'cubic' was equivalent
@@ -115,24 +134,27 @@ Summary of important user-visible change
       * the value of an integer type (int8, uint32, etc.) value exceeds
         the maximum for the format specifier.  For '%d', the limit is
         intmax ('int64') and for '%u' it is intmax ('uint64').
 
       * round(x) != x or the value is outside the range allowed by the
         integer format specifier.
 
     There is still one difference:  Matlab switches to '%e' and Octave
-    is currently switching to '%g'.
+    switches to '%g'.
 
  ** The archive family of functions (bzip2, gzip, zip, tar) and their
     unpacking routines (bunzip2, gunzip, unzip, untar, unpack) have
     been recoded.  Excepting unpack, the default is now to place files
     in the same directory as the archive (on unpack) or as the original
     files (on archiving).
 
+ ** Qt and FLTK graphics toolkits now support offscreen rendering on Linux.
+    In other words, print will work even when the figure visibility is "off".
+
  ** Z-order stacking issues with patches, grid lines, and line object
     plot markers for on screen display and printing have all been resolved.
     For 2-D plots the axis grid lines can be placed on top of the plot
     with set (gca, "layer", "top").
 
  ** The patch graphic object has been overhauled.  It now produces visual
     results equivalent to Matlab even for esoteric combinations of
     faces/vertices/cdata.
@@ -154,40 +176,57 @@ Summary of important user-visible change
  ** The following functions now support N-dimensional arrays:
 
       fliplr   flipud   rot90
 
  ** The new warning ID "Octave:data-file-in-path" replaces the three
     previous separate warning IDs "Octave:fopen-file-in-path",
     "Octave:load-file-in-path", and "Octave:md5sum-file-in-path".
 
+ ** The warning ID Octave:singular-matrix-div has been replaced by
+    Octave:nearly-singular-matrix and Octave:singular-matrix.
+
+ ** The warning ID Octave:matlab-incompatible has been replaced by
+    Octave:language-extension to better reflect its meaning.
+
+ ** The warning ID Octave:broadcast has been removed.  Instead automatic
+    broadcasting will throw an Octave:language-extension warning.  This
+    warning ID is used for broadcasting as well as other features not
+    available in Matlab.
+
  ** Other new functions added in 4.0:
 
       annotation
       bandwidth
       cubehelix
       dir_in_loadpath
       flip
       frame2im
       hgload
       hgsave
       ichol
       ilu
       im2frame
       isbanded
       isdiag
+      isstudent
       istril
       istriu
       javachk
+      jit_failcnt
       linkaxes
       lscov
+      metaclass
       numfields
+      open
       ordschur
+      pan
       qmr
       rotate
+      rotate3d
       sylvester
       unsetenv
       validateattributes
       zoom
 
  ** inline() scheduled for eventual deprecation by Matlab
 
     Functions created through the use of inline are scheduled for deprecation
@@ -212,16 +251,17 @@ Summary of important user-visible change
       fnmatch              | glob or regexp
       loadaudio            | audioread
       luinc                | ilu or ichol
       mouse_wheel_zoom     | mousewheelzoom axes property
       nfields              | numfields
       octave_tmp_file_name | tempname
       playaudio            | audioplayer
       saveaudio            | audiowrite
+      strmatch             | strncmp or strcmp or regexp
       syl                  | sylvester
       usage                | print_usage
 
       allow_noninteger_range_as_index
       do_braindead_shortcircuit_evaluation
       setaudio
 
  ** The following functions were deprecated in Octave 3.8 and will be
@@ -260,36 +300,23 @@ Summary of important user-visible change
     in Octave 4.0 and will be removed from Octave 4.4 (or whatever version is
     the second major release after 4.0).  Use the Fatan2 function with sparse
     inputs as a replacement.
 
  ** The internal class <Octave_map> was deprecated in Octave 3.8 and has
     been removed from Octave 4.0.  Replacement classes are
     <octave_map> (struct array) or <octave_scalar_map> for a single structure.
 
- ** The warning ID Octave:singular-matrix-div has been replaced by
-    Octave:nearly-singular-matrix and Octave:singular-matrix.
-
- ** The warning ID Octave:matlab-incompatible has been replaced by
-    Octave:language-extension to better reflect its meaning.
-
- ** The warning ID Octave:broadcast has been removed.  Instead automatic
-    broadcasting will throw an Octave:language-extension warning, together
-    with other features not available in Matlab.
-
  ** The configure option --enable-octave-allocator has been removed.
     The internal class <octave_allocator> declared in oct-alloc.h has
     been removed.  The header remains, but is deprecated.  The macros to
     access the class (DECLARE_OCTAVE_ALLOCATOR, DEFINE_OCTAVE_ALLOCATOR,
     and DEFINE_OCTAVE_ALLOCATOR2) are now unconditionally defined to be
     empty.
 
- ** Qt and FLTK graphics toolkit now support offscreen rendering,
-    i.e., print will work when figure visibility is off.
-
  ** Octave now has OpenMP enabled by default if the system provides a
     working OpenMP implementation.  This allows oct-file modules to take
     advantage of OpenMP if desired.  This can be disabled when building
     Octave with the configure option --disable-openmp.
 
 ---------------------------------------------------------
 
 See NEWS.3 for old news.
